api/client/build.go

```diff
@@ -390,36 +278,209 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 	return err
 }
 
-var (
-	dockerfileFromLinePattern        = regexp.MustCompile(`(?i)^FROM[ \f\r\t\v]+(?P<image>[^ \f\r\t\v\n#]+)`)
-	dockerfileLineContinutionPattern = regexp.MustCompile(`\\[ \t]*$`)
-)
+func getDockerfileRelPath(dockerfileName *string, baseDir string) (err error) {
+	origDockerfile := *dockerfileName // For log output.
+
+	absBase, err := filepath.Abs(baseDir)
+	if err != nil {
+		return fmt.Errorf("unable to get absolute context directory: %v", err)
+	}
+
+	// If not already an absolute path, the Dockerfile path should be relative
+	// to the base directory.
+	dockerfileAbspath := *dockerfileName
+	if !filepath.IsAbs(dockerfileAbspath) {
+		dockerfileAbspath = filepath.Join(absBase, *dockerfileName)
+	}
+
+	if *dockerfileName == "" {
+		// No -f/--file was specified so use the default
+		dockerfileAbspath = filepath.Join(absBase, api.DefaultDockerfileName)
+
+		// Just to be nice ;-) look for 'dockerfile' too but only
+		// use it if we found it, otherwise ignore this check
+		if _, err = os.Lstat(dockerfileAbspath); os.IsNotExist(err) {
+			altPath := filepath.Join(absBase, strings.ToLower(api.DefaultDockerfileName))
+			if _, err = os.Lstat(altPath); err == nil {
+				dockerfileAbspath = altPath
+			}
+		}
+	}
+
+	// Verify that 'filename' is within the build context
+	dockerfileAbspath, err = symlink.FollowSymlinkInScope(dockerfileAbspath, absBase)
+	if err != nil {
+		return fmt.Errorf("The Dockerfile (%s) must be within the build context (%s)", origDockerfile, baseDir)
+	}
+
+	if _, err := os.Lstat(dockerfileAbspath); err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("Cannot locate Dockerfile: baseDir: %q, origDockerfile: %q, absDockerfile: %q", baseDir, origDockerfile, dockerfileAbspath)
+		}
+		return fmt.Errorf("unable to stat Dockerfile: %v", err)
+	}
+
+	if *dockerfileName, err = filepath.Rel(absBase, dockerfileAbspath); err != nil {
+		return fmt.Errorf("unable to get relative Dockerfile path: %v", err)
+	}
+
+	return nil
+}
+
+func createDockerfileFromReader(r io.Reader, filename string) error {
```

Comment:
This should be called `writeFile or something like that. Alternatively, this can be completely replaced with `ioutil.WriteFile`.